### ISSUE TITLE
feat(weapons-ai): implement SOTA fine-tuning pipeline and negative mining protocol

### ISSUE THREAD

**Issue by mauriziokeningen** [mauriziokeningen/Computer-Vision-Residential-Security-System#36]
# feat(weapons-ai): implement SOTA fine-tuning pipeline and negative mining protocol

## Executive Summary & Context
Our current End-to-End (E2E) pipeline utilizes a baseline YOLOv10s model functioning as a tracer bullet. However, the existing weights were trained on generic public datasets, resulting in severe **Domain Shift**. This manifests as high false-positive rates (hallucinations) and significant degradation in detecting small/distant objects typical in CCTV environments.

This issue tracks the deployment of a production-grade **Transfer Learning** pipeline. By freezing the pre-trained backbone and fine-tuning the detection head with a custom, high-resolution dataset, we will achieve a zero-hallucination tolerance for lethal threat detection.

## Phase 1: Data Architecture & Negative Mining
The success of this fine-tuning iteration relies strictly on the data composition. The dataset MUST be curated to reflect the production environment:

1. **Negative Mining (Anti-Hallucination):** 10% to 20% of the training dataset MUST consist of "Background Images" (e.g., empty hallways, residents holding phones/keys). These images must be included in the dataset WITHOUT corresponding `.txt` annotation files to actively penalize false positives.
2. **Domain-Specific Constraints:** Images must reflect actual CCTV deployment parameters (high pitch angle, small object scale). 
3. **Data Augmentation Guardrails:** Destructive augmentations (`mosaic`, `mixup`, extreme `scale`) are strictly restricted to protect the spatial integrity of sub-30px distant objects.

## Phase 2: Execution Protocol & Recommended Baseline

**Engineering Note (A/B Testing Welcome):** The `train_cctv_weapons.py` script below represents the *recommended SOTA baseline* for CCTV architectures based on current heuristics. However, the assigned AI Engineer is encouraged to challenge these parameters. If empirical A/B testing yields a superior hyperparameter matrix (validated via confusion matrices and `mAP@50-95`), the script should be updated accordingly.

### Architectural Parameter Justifications (For Reviewers):
* **`imgsz=1280`:** CCTV feeds are high definition. Downscaling to 640px crushes distant weapons into unreadable 4x4 pixel artifacts, destroying convolutional geometry. We trade batch size for vital spatial resolution.
* **`mosaic=0.1` & `mixup=0.0`:** Weapons are solid, rigid objects. Heavy mosaic or transparency (mixup) forces the model to learn impossible "ghost" artifacts, destroying accuracy for small targets.
* **`cls=1.5`:** Heavily penalizes the network for confusing a knife with a gun, forcing strict class boundaries.

```python
"""
SOTA Object Detection Fine-Tuning Pipeline (CCTV Domain)
Architecture: YOLOv10s Backbone
Target Hardware: NVIDIA RTX 4070 Ti Super (16GB VRAM)
"""
import os
import torch
import multiprocessing
from ultralytics import YOLO
import logging

logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger("TrainingPipeline")

def optimize_hardware_state():
    if torch.cuda.is_available():
        torch.cuda.empty_cache()
        torch.backends.cudnn.benchmark = True
        logger.info(f"Hardware optimized. Detected GPU: {torch.cuda.get_device_name(0)}")
    else:
        logger.warning("CUDA not detected. Training on CPU will be highly inefficient.")

def execute_fine_tuning():
    logger.info("Initializing Backbone from yolov10s.pt...")
    model = YOLO("yolov10s.pt")

    training_args = {
        "data": "data.yaml",
        "epochs": 150,
        "patience": 25,
        
        # Spatial Retention for Small Objects
        "imgsz": 1280,
        "batch": 16,
        "device": 0,
        "workers": 8,
        "amp": True,

        # Optimization & Strict Class Penalty
        "optimizer": "AdamW",
        "lr0": 0.0005,
        "lrf": 0.01,
        "cos_lr": True,
        "weight_decay": 0.01,
        "warmup_epochs": 3,
        "cls": 1.5,
        "box": 7.5,
        "dfl": 1.5,

        # Non-Destructive CCTV Augmentation
        "mosaic": 0.1,
        "mixup": 0.0,
        "scale": 0.2,
        "degrees": 0.0,
        "perspective": 0.001,
        "fliplr": 0.5,
        "hsv_h": 0.015,
        "hsv_s": 0.5,
        "hsv_v": 0.6,

        "save": True,
        "save_period": 10,
        "plots": True
    }

    logger.info("Starting training sequence...")
    results = model.train(**training_args)
    logger.info(f"Training terminated. Production weights located at: {results.save_dir}/weights/best.pt")

if __name__ == "__main__":
    multiprocessing.freeze_support()
    optimize_hardware_state()
    execute_fine_tuning()
```

## Phase 3: Post-Mortem Validation Strategy
Validation is not strictly bound to `mAP@50`. Upon training completion, the assigned engineer must review the `confusion_matrix.png` generated in the run directory. 
* The intersection of `Background` (True) and `Weapon` (Predicted) must reflect a minimal error rate, proving the efficacy of the Negative Mining strategy.
* The intersection of `Knife` and `Gun` must reflect near-zero confusion.

## Definition of Done (DoD)
- [ ] A custom dataset is compiled in Roboflow/local storage, containing at least 15% unannotated background images.
- [ ] The `train_cctv_weapons.py` script executes without VRAM OOM errors (adjust batch size to 8 if necessary).
- [ ] The training completes without catastrophic overfitting (dictated by the `patience` threshold).
- [ ] The confusion matrix validates the successful suppression of hallucinations and class confusion.
- [ ] The new `best.pt` file is successfully hot-swapped into the production inference pipeline.
